### PR TITLE
gt-178 feature: add constant

### PR DESCRIPTION
**Pull request by epcAnker**
adding constant as an interface item.

Related:
PM: https://github.com/altendky/pm/pull/15
grid-tied: epcpower/grid-tied#71
embedded lib : epcpower/embedded-library#24

### CODE REVIEW

**Review by altendky — COMMENTED** [review]
Is read only a property of the SunSpec data point?  Should there be a check on export or such to ensure that constant parameters aren't attached to not-read-only SunSpec data points?  It feels like there should be something somewhere, maybe.

File: epyqlib/attrsmodel.py [hunk 1]
```diff
@@ -735,6 +735,35 @@ def to_decimal_or_none(s):
...
+
+    if isinstance(s, str) and len(s) == 0:
+        return None
+
+    if isinstance(s, str):
+        s = locale.delocalize(s)
+
+    #see if its a decimal
+    try:
+        if s.isdecimal():
```
[the remark this answers is not in the document]
Comment (epcAnker) [reply epyqlib/attrsmodel.py:750]: will be removing this and adding it back in the next story.  Will make some adjustments based on this comment in the next story.

File: epyqlib/attrsmodel.py [hunk 2]
```diff
@@ -735,6 +735,35 @@ def to_decimal_or_none(s):
...
+            return result
+    except decimal.InvalidOperation as e:
+        print ("to_decimal_or_expression_or_none NOT A DECIMAL")
+        #raise ValueError("Invalid number: {}".format(repr(s))) from e
+
+    #expression
+
+    if isinstance(s, str):
+        if len(s) == 0:
+            return None
```
[the remark this answers is not in the document]
Comment (epcAnker) [reply epyqlib/attrsmodel.py:761]: same as above

**Review by altendky — CHANGES_REQUESTED** [review]
There are some test failures.  The first one looks good.  The second one looks like a somewhat annoying test.  If I remember correctly you can take the output and paste it into the sample file then check the diff with git and make sure the changes are what you intended.

https://ci.appveyor.com/project/KyleAltendorf/stlib/builds/39703883#L2890
```
__________________ TestAttrsModel.test_all_fields_in_columns __________________
self = <epyqlib.attrsmodel.BuiltTests object at 0x0000020B0CC45D88>
    def test_all_fields_in_columns(self):
        fields = set()
    
        for cls in self.types:
            if cls is self.root_type:
                continue
    
            for field in epyqlib.attrsmodel.fields(cls):
                if field.no_column:
                    continue
    
                fields.add((cls, field.name))
    
        columns_list = [
            tuple(x)
            for x in itertools.chain(*(column.items() for column in self.columns))
            if x[0] is not self.root_type
        ]
    
        columns = set(columns_list)
    
        assert len(columns_list) == len(columns)
    
        extra = columns - fields
        missing = fields - columns
    
        assert extra == set()
>       assert missing == set(), columns_to_code(missing)
E       AssertionError: 
E         ((Parameter, 'constant'),),
E       assert {(<class 'epyqlib.pm.parametermodel.Parameter'>, 'constant')} == set()
E         Extra items in the left set:
E         (<class 'epyqlib.pm.parametermodel.Parameter'>, 'constant')
E         Full diff:
E         - {(<class 'epyqlib.pm.parametermodel.Parameter'>, 'constant')}
E         + set()
epyqlib\attrsmodel.py:1616: AssertionError
```

**Comment by epcAnker** [comment]
> Is read only a property of the SunSpec data point? Should there be a check on export or such to ensure that constant parameters aren't attached to not-read-only SunSpec data points? It feels like there should be something somewhere, maybe.

test

**Review by altendky — APPROVED** [review]
It might be worth it at some point to add an entry to the example that has a constant specified.

**Pull request merged**